src/js/media-tooltip.ts

```diff
@@ -0,0 +1,239 @@
+ * @cssproperty --media-tooltip-white-space - `white-space` property of tooltip
+ * @cssproperty --media-tooltip-arrow-display - `display` property of tooltip arrow
+ */
+class MediaTooltip extends globalThis.HTMLElement {
+  static get observedAttributes(): string[] {
+    return [Attributes.POSITION, Attributes.CONTAINER];
+  }
+
+  arrowEl: HTMLElement;
+
+  constructor() {
+    super();
+
+    if (!this.shadowRoot) {
+      // Set up the Shadow DOM if not using Declarative Shadow DOM.
+      this.attachShadow({ mode: 'open' });
+      this.shadowRoot.appendChild(template.content.cloneNode(true));
+    }
+
+    this.arrowEl = this.shadowRoot.querySelector('#arrow');
+  }
+
+  // Adjusts tooltip position relative to the closest specified container
+  // such that it doesn't spill out of the left or right sides. Only applies
+  // to 'top' and 'bottom' positioned tooltips.
+  updateXOffset = () => {
+    const position = this.position;
+
+    // we don't offset against tooltips coming out of left and right sides
+    if (position === 'left' || position === 'right') {
+      // could have been offset before switching to a new position
+      this.style.removeProperty('--media-tooltip-offset-x');
+      return;
+    }
+
+    const containingSelector = this.container;
+    const containingEl = closestComposedNode(this, containingSelector);
+    if (!containingEl) return;
+    const { x: containerX, width: containerWidth } =
+      containingEl.getBoundingClientRect();
+    const { x: tooltipX, width: tooltipWidth } = this.getBoundingClientRect();
+    const tooltipRight = tooltipX + tooltipWidth;
+    const containerRight = containerX + containerWidth;
+    const offsetXVal = this.style.getPropertyValue('--media-tooltip-offset-x');
+    const currOffsetX = offsetXVal
+      ? parseFloat(offsetXVal.replace('px', ''))
+      : 0;
+
+    // we might have already offset the tooltip previously so we remove it's
+    // current offset from our calculations
+    const leftDiff = tooltipX - containerX + currOffsetX;
+    const rightDiff = tooltipRight - containerRight + currOffsetX;
+
+    // out of left bounds
+    if (leftDiff < 0) {
+      this.style.setProperty('--media-tooltip-offset-x', `${leftDiff}px`);
+      return;
+    }
+
+    // out of right bounds
+    if (rightDiff > 0) {
+      this.style.setProperty('--media-tooltip-offset-x', `${rightDiff}px`);
+      return;
+    }
+
+    // no spilling out
+    this.style.removeProperty('--media-tooltip-offset-x');
+  };
+
+  /**
+   * Get or set tooltip position
+   */
+  get position(): TooltipPosition | undefined {
+    return getStringAttr(this, Attributes.POSITION);
+  }
+
+  set position(value: TooltipPosition | undefined) {
+    setStringAttr(this, Attributes.POSITION, value);
+  }
+
+  /**
+   * Get or set tooltip container selector that will constrain the tooltips horizontal
+   * position. A css selector that matches one of the tooltips parents.
+   */
+  get container(): string | undefined {
```

Comment:
`container` feels a little vague here, or at least I'd expect an element from it not string. It's hard to think of something better. Maybe `containWithin`. I like that because it's a little clearer what the functionality is and I don't expect any specific values from it, so query selector values seem fine to use.

Reply:
I had some more verbose names for this initially so was hoping someone would have an idea for something descriptive and succinct. I think `containWithin` is pretty good. @cjpillsbury also mentioned that to follow existing patterns we should accept an element as well as a string potentially.

Reply:
Have updated this to `containWithin`

Reply:
> we should accept an element as well as a string potentially

I think that's right in general but this is a situation kind of like the `mediacontroller` attribute where I think it'd be good if we didn't hold on to a reference of the element. Unless we need to do that anyway.

Reply:
if we're sticking with generic class selectors (instead of ids) this should work for any of our "automagic container picking" solutions